internal/utils/temporary_utils.go

```diff
@@ -0,0 +1,118 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func SetAddonInstanceCondition(ctx context.Context, cacheBackedKubeClient client.Client, condition metav1.Condition, addonName string) error {
+	addonInstance, err := getAddonInstanceByAddon(ctx, cacheBackedKubeClient, addonName)
+	if err != nil {
+		return fmt.Errorf("failed to fetch AddonInstance by the addon '%s': %w", addonName, err)
+	}
+	if err := upsertAddonInstanceCondition(ctx, cacheBackedKubeClient, &addonInstance, condition); err != nil {
+		return fmt.Errorf("failed to update the conditions of the AddonInstance resource in the namespace %s: %w", addonInstance.Namespace, err)
+	}
+	return nil
+}
+
+type ReconcilerWithHeartbeat interface {
+	reconcile.Reconciler
+	GetLatestHeartbeat() metav1.Condition
+	SetLatestHeartbeat(metav1.Condition)
+}
+
+func SetupHeartbeatReporter(r ReconcilerWithHeartbeat, mgr manager.Manager, addonName string, handleAddonInstanceConfigurationChanges func(addonsv1alpha1.AddonInstanceSpec)) error {
+	defaultHealthyHeartbeatConditionToBeginWith := metav1.Condition{
+		Type:    "addons.managed.openshift.io/Healthy",
+		Status:  "True",
+		Reason:  "AddonHealthy",
+		Message: fmt.Sprintf("Addon '%s' is perfectly healthy", addonName),
+	}
+	// initialized with a healthy heartbeat condition corresponding to the addon
+	r.SetLatestHeartbeat(defaultHealthyHeartbeatConditionToBeginWith)
+
+	heartbeatReporterFunction := func(ctx context.Context) error {
+		currentAddonInstanceConfiguration, err := GetAddonInstanceConfiguration(ctx, mgr.GetClient(), addonName)
+		if err != nil {
+			return fmt.Errorf("failed to get the AddonInstance configuration corresponding to the Addon '%s': %w", addonName, err)
+		}
+
+		// Heartbeat reporter section: report a heartbeat at an interval ('currentAddonInstanceConfiguration.HeartbeatUpdatePeriod' seconds)
+		for {
+			currentHeartbeatCondition := r.GetLatestHeartbeat()
+			if err := SetAddonInstanceCondition(ctx, mgr.GetClient(), currentHeartbeatCondition, addonName); err != nil {
+				mgr.GetLogger().Error(err, "error occurred while setting the condition", "HeartbeatCondition", fmt.Sprintf("%+v", currentHeartbeatCondition))
+			}
+
+			// checking latest addonInstance configuration and seeing if it differs with current AddonInstance configuration
+			latestAddonInstanceConfiguration, err := GetAddonInstanceConfiguration(ctx, mgr.GetClient(), addonName)
+			if err != nil {
+				return fmt.Errorf("failed to get the AddonInstance configuration corresponding to the Addon '%s': %w", addonName, err)
+			}
+			if !reflect.DeepEqual(currentAddonInstanceConfiguration, latestAddonInstanceConfiguration) {
+				currentAddonInstanceConfiguration = latestAddonInstanceConfiguration
+				handleAddonInstanceConfigurationChanges(currentAddonInstanceConfiguration)
+			}
+
+			// waiting for heartbeat update period for executing the next iteration
+			<-time.After(currentAddonInstanceConfiguration.HeartbeatUpdatePeriod.Duration)
```

Comment:
We need to change this up a bit, when an Addon is reporting unhealthy or healthy, this status update should go through immediately and not be blocked by the next heartbeat timeout.

The HeartbeatUpdatePeriod is there to update "at least as often", it's not meant as a limit to delay status updates in between.

Reply:
I guess a straightforward approach to do this is, modify the `SetLatestHeartbeat(metav1.Condition)` method to
```go
func (r *ReferenceAddonReconciler) SetLatestHeartbeat(heartbeat metav1.Condition) {
	r.latestHeartbeat = heartbeat
        addonName := "reference-addon" // can be also automatically picked by Downwards API from addon's Deployment resource?
        SetAddonInstanceCondition(ctx, r.Client, r,GetLatestHeartbeat(), addonName) // error handling to be added to
}
```

Reply:
Although this ends up relying on addon developer's responsibility to do so. Let me think if there's some simple solution to abstract away this real-time heartbeat reporting under the central-heartbeat-reporter-loop itself.